### ISSUE TITLE
Does Viva Wallet (Native Checkout v2 3DS) Work with Woocommerce or it is beta version?

### ISSUE THREAD

**Issue by rohitgilbile7** [VivaPayments/API#1247]
Hello,

I am trying to build a Woocomerce plugin for (Native Checkout v2 3DS). I read many articles and viva wallet documentation but didn't found practical examples to implement Native Checkout v2 3DS.

[Viva Wallet](https://github.com/VivaPayments/API/wiki)) says Native checkout v2 3DS in beta version. Is it or document is not updated?

Has anyone implemented Native checkout v2? Please guide.

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7 just to let you know that Native Checkout v2 has been out of beta for some time. The page on GitHub you mention needed updating which I've done now. The extent of the documentation we have for Native Checkout v2 is listed below:

- [Native Checkout v2](https://developer.vivawallet.com/online-checkouts/native-checkout-v2/) (Viva Wallet Developer Portal page)
- [Migrating from Native Checkout v1](https://developer.vivawallet.com/online-checkouts/migrating-from-native-checkout-v1/)
- [Sample .NET project](https://github.com/VivaPayments/API/tree/master/NativeCheckoutV2/NativeCheckoutV2)

PHP code samples to follow in the near future. 

Hope this helps!

**Comment by rohitgilbile7** [comment]
Hi,

Thanks for the update.

I follow the sample .Net Project to implement in PHP.

And I am Facing CORS policy blocking issue. I thought it is blocking by my server. But when I contact my server regarding the same they replied " CORS policy is blocking access to demo.vivapayments.com, which is a third-party issue. I would highly recommend reaching out to vivapayments.com for further assistance on how to get access to their site.  "

![errr](https://user-images.githubusercontent.com/11695482/69980712-c84bc100-1556-11ea-95b3-548a0c2d2191.png)
![error1](https://user-images.githubusercontent.com/11695482/69980711-c84bc100-1556-11ea-8cab-4d51c213bcfc.png)

https://admitcardjobs.in/havanahouse/payment.php

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7 I'm sorry, but where are you getting the `https://demo.vivapayments.com/3ds/services/nativecheckout/v2/chargetokens` endpoint from? It's not in our documentation or code sample. Correct endpoint is `https://demo-api.vivapayments.com/nativecheckout/v2/chargetokens`

**Comment by rohitgilbile7** [comment]
Hello,

Thanks for the update.
I was following [.NET Project](https://github.com/VivaPayments/API/blob/master/NativeCheckoutV2/NativeCheckoutV2/NativeCheckoutV2/wwwroot/js/site.js) and also [this](https://developer.vivawallet.com/online-checkouts/native-checkout-v2/) to create my Native Checkout in PHP.

I completed my 2 steps in setup instructions. In step 3 to Initialize Native Checkout. I needed **authToken**. I got auth token reference from [here.](https://developer.vivawallet.com/authentication/#identityserver-oauth-2-authentication)

When I make the POST request to **https://demo-accounts.vivapayments.com/connect/token** from the postman I got the access_token.
![postman](https://user-images.githubusercontent.com/11695482/70081980-7fb40680-162f-11ea-8f9b-ba5b1c984489.png)

But when I make the request from AJAX I am getting below error.
![AJAX](https://user-images.githubusercontent.com/11695482/70082222-e9341500-162f-11ea-8962-60654708f93c.png).

Here is my full source code. I took an example from .NET project and documentation.

`<html>
    <head>
        <script src="https://ajax.aspnetcdn.com/ajax/jQuery/jquery-3.4.1.min.js"></script>
        
        <script src="https://www.vivapayments.com/web/checkout/v2/js"></script>

    </head>
    <body>
 	
<div class="row">
    <div class="col-md-10">
        <form id="native-checkout-form" method="post"  >
             <div class="form-group">
                <label>Cardholder's name</label>
                <input data-vp="cardholder" class="form-control js-cardholder">
             </div>

            <div class="row">
                <div class="col-md-4">
                    <div class="form-group">
                        <label>Card number</label>
                        <input data-vp="cardnumber" class="js-card-number form-control" maxlength="19" value="4111111111111111">
                    </div>
                </div>

                <div class="col-md-1">
                    <div class="form-group">
                        <label>Cvv</label>
                        <input data-vp="cvv" class="form-control js-cvv" maxlength="4">
                    </div>
                </div>

                <div class="col-md-2">
                    <div class="form-group">
                        <label>Expires</label>
                        <div class="row"> 
                            <div class="col-md-6" style="padding-right: 5px;">
                                <input data-vp="month" class="js-expiration-month form-control" maxlength="2" placeholder="month">
                            </div>
                            <div class="col-md-6" style="padding-left: 0px;">
                                <input data-vp="year" class="js-expiration-year form-control" maxlength="4" placeholder="year">
                            </div>
                        </div>
                    </div>
                </div>

                <div class="col-md-1">
                    <div class="form-group js-installments" style="display: none;">
                        <label>Installments</label>
                        <select class="form-control">
                            <option selected>1</option>
                        </select>
                    </div>
                </div>
             </div>

            <input class="js-charge-token" type="hidden" name="chargetoken" value=""/>

            <button type="button" class="btn btn-primary js-btn-chargecard">Charge card</button>
        </form>
    </div>
</div>
 <div class="row">
    <div id="threed-pane" class="col-md-12">

    </div>
</div>

<div class="row">
    <div class="js-charge-result alert" role="alert" style="display:none;">

    </div>
</div>
<script>
 $(function () {
  
$.ajax({
    url : "https://demo-accounts.vivapayments.com/connect/token",
    type: "POST",
   /* beforeSend: function (xhr){ 
       xhr.setRequestHeader('Authorization', 'Basic eGNpd2Nhbjg3OHQ3bjAwM3prdDVvMHJ1dTVnMmMyd2NhNTh1OWo2dDJhaDQzLmFwcHMudml2YXBheW1lbnRzLmNvbTpwWjgxN2pZVzlvVkE5OHhPVFE0M1VqUzlDOExpME0='); 
     }, */
    data :  {
        grant_type: 'client_credentials',
       client_id: 'xciwcan878t7n003zkt5o0ruu5g2c2wca58u9j6t2ah43.apps.vivapayments.com',
        client_secret:'pZ817jYW9oVA98xOTQ43UjS9C8Li0M'
    },

    success: function(data, textStatus, jqXHR)
    {
     var  accessToken =  data.access_token;
            console.log(data);
                alert('Thanks for your comment!');     },
    error: function (jqXHR, textStatus, errorThrown)
    {
        alert('error');
                         console.log(jqXHR.responseText);
    }
}); 
        
    if (!window.accessToken) {
        alert('No access token is available');
    }
    $('.js-btn-chargecard').on('click', function () {
        
         
        let options = {
            installments: 1
        };
        let installments = 1 //$('.js-installments select').val();

        if (installments >= 1) {
            options.installments = installments;
        }
        options.amount = 22;
              options.authenticateCardholder  = false

        console.log(options);
        VivaPayments.cards.requestToken(options)
            .done(function (data) {
                if (data.Error) {
                    alert('if');
                    console.log(data.Error);
                    return;
                }
                 let paymentForm = $('#native-checkout-form');
                alert(data.chargeToken);
                $('.js-charge-token').val(data.chargeToken);

                $.ajax({
                    url: 'http://admitcardjobs.in/havanahouse/payment.php',
                    type: 'POST',
                    data: paymentForm.serialize()
                }).done(function (responseData) {
                    if (responseData && responseData.statusId === 'F') {
                        $('.js-charge-result').html(
                            `Charge was successful. Transaction Id ${responseData.transactionId}`)
                            .addClass('alert-success')
                            .show();
                    }
                }).fail(function (xhrData) {
                    $('.js-charge-result')
                        .html('Card could not be charged')
                        .addClass('alert-danger')
                        .show();
                    });
            }).fail(function (responseData) {
                    alert('Here is the reason it failed: '  + responseData.Error.toString());
                  console.log('Here is the reason it failed: '  + responseData.Error.toString());
                  console.log(responseData);
         });
    });
});
    
</script>
        </body>
</html>`

Can you please help me to implement Native checkout V2 3Ds. I follow the documentation and .NET project but facing this weird issues.

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7, thanks for sending through further info. However, I need to ask again, where are you getting the https://demo.vivapayments.com/3ds/services/nativecheckout/v2/chargetokens endpoint from? As far as I'm aware `3ds/services` is incorrect in this endpoint!

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7 response from our development team:
>In order to request an access token you must provide your `clientId` and `clientSecret` which are sensitive data and should not be exposed in the browser.
This call must be made from your back-end.

**Comment by rohitgilbile7** [comment]
> Hi @rohitgilbile7, thanks for sending through further info. However, I need to ask again, where are you getting the https://demo.vivapayments.com/3ds/services/nativecheckout/v2/chargetokens endpoint from? As far as I'm aware `3ds/services` is incorrect in this endpoint!

As I said I was following your. NET project. I tried to create a clone in PHP. And I got that 

> 3ds/services
error.

**Comment by matthewrgourd** [comment]
@rohitgilbile7 My apologies, I can see that now...let me check again with our development team.

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7 I'm spoken to our development team and apparently that .NET project I mentioned is out of date which explains the incorrect URL. Sorry for this! However, you're also trying to generate an access_token from ([JavaScript](https://pastebin.com/rQyAu2Tf) line: 79) which renders in the client browser. This call must be made from your back-end server (PHP). Hope this helps!

**Comment by rohitgilbile7** [comment]
Thanks for the update. Will try to generate access  token from backend. And will update you soon

**Comment by matthewrgourd** [comment]
We will look into this further in response to your developers' comments below:

>According to viva wallet documentation I had setup the code and finished to most of the task.
According to viva wallet documentation before sending users data (Card number, CVV). You need "authToken". And we can get this by calling their API. But when I make API call getting response CORS policy blocked issue.

**Comment by matthewrgourd** [comment]
Hi @rohitgilbile7 is this still a problem? If so., please send your full source code to mgourd@vivawallet.com referencing this issue and I'll get it checked out by a developer. Otherwise please close this ticket.

**Comment by rohitgilbile7** [comment]
Hello,
I am getting an empty response when I make a request from my backend to make an actual charge.
When I print HTTP request code in a response I receive 401.
Below is my sample full source code. 

```
<?php
      error_reporting(E_ALL); 
       ini_set('display_errors', 1);
 	if(isset($_POST['chargetoken'])){
 	    $chargeToken = $_POST['chargetoken'];
 	    $curl = curl_init();
 	// Set some options - we are passing in a useragent too here

        $headers = array(
             'Content-Type: application/json',
             'Authorization: Bearer eGNpd2Nhbjg3OHQ3bjAwM3prdDVvMHJ1dTVnMmMyd2NhNTh1OWo2dDJhaDQzLmFwcHMudml2YXBheW1lbnRzLmNvbTpBOHk3NTV1YVdYVTBmQmFnUEtlVXBWaDNNTllBQlUK');
        curl_setopt_array($curl, [
            CURLOPT_RETURNTRANSFER => 1,
            CURLOPT_URL => 'https://demo-api.vivapayments.com/nativecheckout/v2/transactions',
            CURLOPT_USERAGENT =>  $_SERVER['HTTP_USER_AGENT'],
            curl_setopt($curl, CURLOPT_HTTPHEADER, $headers),
            CURLOPT_POST => 1,
            CURLOPT_POSTFIELDS => [
                 'amount'=> 100,
                  'preauth'=> false,
                  'sourceCode'=> "4430",
                  'chargeToken'=> $chargeToken,
                  'Installments'=> '1',
                  'merchantTrns'=> "Merchant transaction reference",
                  'customerTrns'=> "Description that the customer sees",
                  'currencyCode'=> '826',
                // item1 => 'value',
                // item2 => 'value2'
            ]
        ]);
        // Send the request & save response to $resp

        $resp = curl_exec($curl);
        $httpCode = curl_getinfo($curl , CURLINFO_HTTP_CODE); // this results 0 every time

        // Close request to clear up some resources
      if ($resp === false) {
          $response = curl_error($curl);
           echo stripslashes($response);
      }else{
            echo $resp;
      }
    
 	   // echo $httpCode;
 	             curl_close($curl);

 	}
?>
```

For authorization my base 64 encoding client ID & secret key.
Also Can you please let me know the transaction URL to make the actual charge. Is this correct 

> https://demo-api.vivapayments.com/nativecheckout/v2/transactions
